### ISSUE TITLE
[BUG] Matching energies to reference after dynamics

### ISSUE THREAD

**Issue by alecbetancourt** [deepmodeling/DMFF#121]
### Bug summary

I'm attempting to compare DMFF simulation results in MD runs to some other libraries with AMBER style protein force fields as a starting point for future research into auto differentiable MD tools. I'm utilizing JAX MD as a simulation engine and DMFF for the neighbor list generation and energy computation. For very small molecules (a few dozen atoms), I'm seeing less than ideal agreement with other tools but still seeing good numerical stability for longer simulations. For larger molecules, especially in solvent (thousands of atoms), I'm seeing mixed results in initial agreement for energies and no long term stability for simulations. Although I would expect some of these differences from other packages, it seems like the initial energies should be closer than they are and in a component wise analysis, are the most different in the non bonded forces. I've noticed that varying the neighbor cutoff changes these results considerably and have been using simulation parameters that mimic the defaults found in most simple AMBER runs (8 angstrom, time step choice, etc). I've tried exploring a few different avenues for things I think may be causing these issues, but I've been unable to notice any trends in these irregularities, even when comparing statistics for energies/forces on a per atom basis.

Although I'm not entirely sure if there are some issues compounding to cause this or user error, I've more generally been looking to see if there good examples of using DMFF energy/force generation in traditional point charge model MD simulations as opposed to just instantaneous energies/forces.

Using an example of a 1289 atom system for the RAMP1 protein and comparing to OpenMM:
Initial Pot E:  -1550.4843600621414 kJ/mol
"Step","Potential Energy (kJ/mole)"
100,-8029.34375
200,-8204.943359375
300,-8464.1376953125
400,-8592.5234375
500,-8519.421875
600,-8564.0693359375
700,-8481.87890625
800,-8524.376953125
900,-8626.42578125
1000,-8622.3681640625

Assuming units are in nanometers, comparing this to DMFF with what should be an 8 angstrom cutoff (.8 nanometers in the code for internal consistency) yields fairly different initial energies and results that quickly become unstable.
Initial Potential Energy
3388.2755606199653
Statistics after timestep 100
Potential Energy 7426.3045792548555
Statistics after timestep 200
Potential Energy 6702.456583081994
Statistics after timestep 300
Potential Energy 6629.48060030915
Statistics after timestep 400
Potential Energy 360799615965.03326

Comparing this to another DMFF run where the neighbor list cutoff is increased to 8 (this should be 8 nanometers, not angstrom from my understanding), it seems like there's better initial agreement but instantaneous instability.
Initial Potential Energy
1320.4532334809635
Statistics after timestep 1
Potential Energy 3619221529.320178

### DMFF Version

0.2.1

### JAX Version

jaxlib[cuda11_cudnn805]==0.3.15

### OpenMM Version

7.7.0

### How did you download the software?

pip

### Input Files, Running Commands, Error Log, etc.

My attempt at getting a working MD simulation with DMFF energies:
```
import sys
import time
import jax
import jax.numpy as jnp
import numpy as np
import openmm.app as app
import openmm.unit as unit
from dmff import Hamiltonian, NeighborList, NeighborListFreud

from jax_md import space, smap, energy, minimize, quantity, simulate, quantity

from jax.config import config
config.update("jax_enable_x64", True)

prmtop = app.AmberPrmtopFile('../RAMP1_gas.prmtop')
inpcrd = app.AmberInpcrdFile('../RAMP1_gas.inpcrd')
ff = Hamiltonian("amber14/protein.ff14SB.xml", "amber14/tip3p.xml")

potentials = ff.createPotential(prmtop.topology, nonbondedMethod=app.NoCutoff)
params = ff.getParameters()
positions = jnp.array(inpcrd.getPositions(asNumpy=True).value_in_unit(unit.nanometer))

box = jnp.array([
    [20.0, 0.0, 0.0], 
    [0.0, 20.0, 0.0],
    [0.0, 0.0, 20.0]
])

nbList = NeighborListFreud(box, 8, potentials.meta["cov_map"])
nbList.allocate(positions)
efunc = potentials.getPotentialFunc()
masses = jnp.array([a.element.mass.value_in_unit(unit.dalton) for a in prmtop.topology.atoms()]).reshape((-1, 1))
displacement_fn, shift_fn = space.periodic_general(box, fractional_coordinates=False)
key = jax.random.PRNGKey(0)
energy_fn = lambda pos, **kwargs: efunc(pos, box, kwargs["pairs"], params)
init_fn, apply_fn = simulate.nve(energy_fn, shift_fn, 1e-3)
state = init_fn(key, positions, mass=masses, pairs=nbList.pairs, kT=0)

def body_fn(i, stateList):
  state, nbList, trimpairs = stateList
  nbList.update(state.position)
  trimpairs = nbList.pairs
  state = apply_fn(state, pairs=trimpairs)
  return state, nbList, trimpairs

print("Initial Energy w/ JAX State")
print(energy_fn(state.position, pairs=nbList.pairs)/4.184)

trimmedpair = nbList.pairs
for i in range(10):
  print("Statistics after timestep", (i+1)*100)
  for j in range(100):
    state, jnbList, trimmedpair = body_fn(j, (state, nbList, trimmedpair))
  print("Potential Energy ", end="")
  nrg = energy_fn(state.position, pairs=trimmedpair)
  print(nrg)
```

Attempting to do something approximately similar with OpenMM:
```
import parmed as pmd
from openmm.app import *
from openmm import *
from parmed.unit import *
from sys import stdout
import time

import inspect

prmtop = AmberPrmtopFile('../RAMP1_gas.prmtop')
inpcrd = AmberInpcrdFile('../RAMP1_gas.inpcrd')
system = prmtop.createSystem(nonbondedMethod=NoCutoff, constraints=None)

integrator = VerletIntegrator(0.001*picoseconds)
platform = Platform.getPlatformByName('CUDA')
simulation = Simulation(prmtop.topology, system, integrator, platform)
simulation.context.setPositions(inpcrd.positions)

state = simulation.context.getState(getEnergy=True, getVelocities=True, getForces=True, getPositions=True)
energy = state.getPotentialEnergy()
print("Initial Pot E: ", energy/4.184)
simulation.reporters.append(StateDataReporter(stdout, 100, step=True,
        potentialEnergy=True, temperature=False))
simulation.step(1000)
```

### Steps to Reproduce

Running the above scripts yields the aforementioned results. Based on some of what I've seen with neighbor list configuration, I'm not sure if I'm making fundamental mistakes in my simulation setup or if there's something else going on in terms of unit conversion/more fundamental issues that are causing the results I'm seeing. Any guidance here would be appreciated!

### Further Information, Files, and Links

Included is a zip of both a smaller, stable molecule I've been testing with (22 atoms - Alanine Dipeptide) and a larger unsolvated molecule with unstable behavior (1289 atoms - RAMP1). Both were prepared with AMBER tools.
[ramp1.zip](https://github.com/deepmodeling/DMFF/files/12911644/ramp1.zip)

**Comment by WangXinyan940** [comment]
Hello. Maybe you would like to switch to branch wangxy/v1.0.0-devel and try this test code:

```
import sys
import time
import jax
import jax.numpy as jnp
import numpy as np
import openmm.app as app
import openmm.unit as unit
import openmm as mm
from dmff import Hamiltonian, NeighborListFreud

from jax.config import config
config.update("jax_enable_x64", True)

prmtop = app.AmberPrmtopFile("RAMP1_gas.prmtop")
inpcrd = app.AmberInpcrdFile("RAMP1_gas.inpcrd")
ff = Hamiltonian("amber14/protein.ff14SB.xml", "amber14/tip3p.xml")

potentials = ff.createPotential(prmtop.topology, nonbondedMethod=app.NoCutoff, useDispersionCorrection=False)
params = ff.getParameters()
positions = jnp.array(inpcrd.getPositions(asNumpy=True).value_in_unit(unit.nanometer))

box = jnp.array([[100.0, 0.0, 0.0], [0.0, 100.0, 0.0], [0.0, 0.0, 100.0]])

nbList = NeighborListFreud(box, 20.0, potentials.meta["cov_map"])
nbList.allocate(positions)
pairs = nbList.pairs
efunc = potentials.getPotentialFunc()
dmff_e = efunc(positions, box, pairs, ff.paramset)
print("DMFF Pot E: ", dmff_e)

system = prmtop.createSystem(nonbondedMethod=app.NoCutoff, constraints=None)
integrator = mm.VerletIntegrator(0.001*unit.picoseconds)
platform = mm.Platform.getPlatformByName('Reference')
simulation = app.Simulation(prmtop.topology, system, integrator, platform)
simulation.context.setPositions(inpcrd.positions)

state = simulation.context.getState(getEnergy=True, getVelocities=True, getForces=True, getPositions=True)
energy = state.getPotentialEnergy()
print("OpenMM Pot E: ", energy)
```

The output on my laptop is:
```
DMFF Pot E:  -6487.265980775686
Initial Pot E:  -6487.229977725502 kJ/mol
```

Our neighbor list is a (N, 3) matrix with the third axis saving covalant distance (1-2, 1-3, 1-4, etc). Directly using neighborlist in jax_md would cause unexpected error.

**Comment by WangXinyan940** [comment]
The devel branch has been merged to the master branch and our 1.0.0 version is released. I believe this problem is solved and would close this issue.
